File: 3152-maximum-value-of-an-ordered-triplet-ii/maximum-value-of-an-ordered-triplet-ii.go
func maximumTripletValue(nums []int) int64 {
    n := len(nums);
    L := int64(nums[0]);
    maxD := L - int64(nums[1]);
    result := max(0, maxD * int64(nums[2]));

    for i := 3; i < n; i++ {
        L = max(L, int64(nums[i - 2]));
        maxD = max(maxD, L - int64(nums[i - 1]));
        result = max(result, maxD * int64(nums[i]));
    }

    return result;
}

func max(a, b int64) int64 {
    if a < b {
        return b;
    }

    return a;
}

File: 3152-maximum-value-of-an-ordered-triplet-ii/maximum-value-of-an-ordered-triplet-ii.java
class Solution {
    public long maximumTripletValue(int[] nums) {
        int n = nums.length;
        long L = nums[0], maxD = L - nums[1], result = Math.max(0L, maxD * nums[2]);

        for (int i = 3; i < n; i++) {
            L = Math.max(L, (long)nums[i - 2]);
            maxD = Math.max(maxD, L - nums[i - 1]);
            result = Math.max(result, maxD * nums[i]);
        }

        return result;
    }
}